### PR TITLE
Improve `BeamRelevant` reduced diagnostics using `AMReX_Tuple`.

### PR DESCRIPTION
**Pull request by Yin-YinjianZhao**
- iterate particles only twice @Yin-YinjianZhao 
- to global MPI communications only twice @ax3l 

For example, https://github.com/ECP-WarpX/WarpX/blob/3fa7d28987e8f641a24f658dc06d4a75eeba3597/Source/Diagnostics/ReducedDiags/BeamRelevant.cpp#L178-L231 are modified like https://github.com/ECP-WarpX/WarpX/blob/3fa7d28987e8f641a24f658dc06d4a75eeba3597/Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp#L126-L149.

### CODE REVIEW

**Comment by Yin-YinjianZhao** [comment]
@ax3l Something like this?
I tested it using `Examples/Physics_applications/plasma_acceleration/inputs_3d_boost` and it produced exactly the same results as before.

File: Source/Diagnostics/ReducedDiags/BeamRelevant.cpp [hunk 1]
```diff
@@ -153,271 +154,165 @@ void BeamRelevant::ComputeDiags (int step)
...
+        ParticleReal y_mean  = amrex::get<2>(r);
+        ParticleReal z_mean  = amrex::get<3>(r);
+        ParticleReal ux_mean = amrex::get<4>(r);
+        ParticleReal uy_mean = amrex::get<5>(r);
+        ParticleReal uz_mean = amrex::get<6>(r);
+        ParticleReal gm_mean = amrex::get<7>(r);
 
         // reduced sum over mpi ranks
-        ParallelDescriptor::ReduceRealSum(x_mean);  x_mean  /= w_sum;
+        ParallelDescriptor::ReduceRealSum(w);
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/BeamRelevant.cpp]: I'll combine those individual MPI calls into one with an extra commit :)

**Comment by ax3l** [comment]
Thanks a lot! I pushed one more update to combine the 2x~10x global MPI communications into two. Can you please review if the results are still correct?

One thing I noticed for a follow-up: we still iterate the particle list twice after this optimization, but everything that needs a mean actually can be calculated in one go.

Check out the math for the sums that go into the std deviations here: https://www.strchr.com/standard_deviation_in_one_pass?allcomments=1

File: Source/Diagnostics/ReducedDiags/BeamRelevant.cpp (hunk 1 above)
Comment (ax3l) [reply Source/Diagnostics/ReducedDiags/BeamRelevant.cpp]: done

**Comment by ax3l** [comment]
Another tiny idea for a follow-up:
In order to simplify post-processing, I would also say we just always write out `y` related quantities as zero (in 2D):
https://warpx.readthedocs.io/en/latest/developers/dimensionality.html#conventions

**Comment by Yin-YinjianZhao** [comment]
@ax3l Some values are only 1 digit different this time. I believe it is okay.
For example,
1.96523733622265e-05
1.96523733622266e-05

For the standard deviation calculation:
This comment says there would be a small error.

```
16 comments
orib, 15 years ago
This trick is mathematically valid; However, you may find yourself suffering from numerical underflow errors.

Let's say each X is either (2**40)+1 or (2**40)-1 with probability 0.5. Two loop computation would yield a mean of 2**40 and correct stddev. One loop computation would produce incorrect stddev results.
Peter Kankowski, 15 years ago
Yes, you are right: sq_sum will be underflowed in one loop computation if a[] consists of large values with low difference between them. Thank you, I did not think about this error.
```
Do you still want to use this trick?

Although `y` values are not zero for 2D, but they are not really outputted at the end.
Do you want to force them to be zero anyway?

**Comment by Yin-YinjianZhao** [comment]
@ax3l Wait, the newest updated version gives no difference.

**Comment by ax3l** [comment]
Yes, there was a shadowing warning that actually was pointing to a bug (mismatch of `w` and `w_sum`)

**Comment by ax3l** [comment]
For the precision of both approaches:
That's correct, but only marginally worse than our current sums. We have a lot of values per GPU, where most small+big number errors are introduces. The additional parallelism from MPI is relatively small compared to this (MPI: up to ~10'000 x parallelism; GPU: 10^6-10^9 particles per GPU that are added).

If we want to improve stability of those, we could in AMReX implement improved reduce kernels that treat roundoff better.
(Some experiments in other AMReX codes were posted here: https://github.com/starkiller-astro/Microphysics/blob/main/util/esum.H)

**Comment by ax3l** [comment]
Oh, found another one for variances that works in one pass and has a stable sum:
https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm

**Comment by Yin-YinjianZhao** [comment]
Tested again using Examples/Physics_applications/plasma_acceleration/inputs_3d_boost and it produced exactly the same results as before.

**Review by ax3l — APPROVED** [review]

**Pull request merged**